certbot/interfaces.py

```diff
@@ -591,3 +599,97 @@ def add_message(self, msg, priority, on_crash=True):
 
     def print_messages(self):
         """Prints messages to the user and clears the message queue."""
+
+
+# Updater interfaces
+#
+# When "certbot renew" is run, Certbot will iterate over each lineage and check
+# if the selected installer for that lineage is a subclass of each updater
+# class. If it is and the update of that type is configured to be run for that
+# lineage, the relevant update function will be called for each domain in the
+# lineage. These functions are never called for other subcommands, so if an
+# installer wants to perform an update during the run or install subcommand, it
+# should do so when :func:`IInstaller.deploy_cert` is called.
+
+class GenericUpdater(object):
+    """Interface for update types not currently specified by Certbot.
+
+    This class allows plugins to perform types of updates that Certbot hasn't
+    defined (yet).
+
+    """
+
+    __metaclass__ = abc.ABCMeta
+
+    @abc.abstractmethod
+    def generic_updates(self, domain, *args, **kwargs):
+        """Perform any update types defined by the installer.
+
+        If an installer is a subclass of the class containing this method, this
+        function will always be called when "certbot renew" is run. If the
+        update defined by the installer should be run conditionally, the
+        installer needs to handle checking the conditions itself.
+
+        This method is called once for each domain.
+
+        :param str domain: domain in the lineage being updated
+
+        """
+
+
+class RenewDeployer(object):
+    """Interface for update types run when a lineage is renewed
+
+    This class allows plugins to perform types of updates that need to run at
+    lineage renewal that Certbot hasn't defined (yet).
+
+    """
+
+    __metaclass__ = abc.ABCMeta
+
+    @abc.abstractmethod
+    def renew_deploy(self, lineage, *args, **kwargs):
```

Comment:
We should give an example here to explain when this should be used -- "for example, if a plugin wants to copy over a certificate to a new location only when it is renewed".

Reply:
Explanation added.